src/graph.rs

```diff
@@ -0,0 +1,221 @@
+        is_free: row.get(5),
+    })?;
+
+    let mut unit_props = Vec::new();
+    for row in rows {
+        unit_props.push(row?);
+    }
+
+    if unit_props.len() != 2 {
+        return Err(format_err!("Not 2 rows for {} and {}", unit1, unit2));
+    }
+
+    let (unit_props1, unit_props2) = if &unit_props[0].unit == unit1 {
+        (&unit_props[0], &unit_props[1])
+    } else {
+        (&unit_props[1], &unit_props[0])
+    };
+
+    compare_unit_props(db, unit_props1, unit_props2)
+}
+
+pub fn compare_unit_props(
+    db: &Connection,
+    unit_props1: &UnitProps,
+    unit_props2: &UnitProps,
+) -> Result<Option<i32>> {
+    if unit_props1.unit == unit_props2.unit {
+        return Ok(Some(0));
+    }
+    if unit_props1.level == unit_props2.level {
+        return Ok(None);
+    }
+    if unit_props1.is_free == 1 && unit_props2.is_free == 1 {
+        // free units
+        return Ok(None);
+    }
+
+    // genesis
+    if unit_props1.latest_included_mc_index == None {
+        return Ok(Some(-1));
+    }
+    if unit_props2.latest_included_mc_index == None {
+        return Ok(Some(1));
+    }
+
+    if unit_props1.latest_included_mc_index >= unit_props2.main_chain_index
+        && unit_props2.main_chain_index != None
+    {
+        return Ok(Some(1));
+    }
+    if unit_props2.latest_included_mc_index >= unit_props1.main_chain_index
+        && unit_props1.main_chain_index != None
+    {
+        return Ok(Some(-1));
+    }
+
+    if unit_props1.level <= unit_props2.level
+        && unit_props1.latest_included_mc_index <= unit_props2.latest_included_mc_index
+        && (unit_props1.main_chain_index <= unit_props2.main_chain_index
+            && unit_props1.main_chain_index != None
+            && unit_props2.main_chain_index != None
+            || unit_props1.main_chain_index == None
+            || unit_props2.main_chain_index == None)
+        || unit_props1.level >= unit_props2.level
+            && unit_props1.latest_included_mc_index >= unit_props2.latest_included_mc_index
+            && (unit_props1.main_chain_index >= unit_props2.main_chain_index
+                && unit_props1.main_chain_index != None
+                && unit_props2.main_chain_index != None
+                || unit_props1.main_chain_index == None
+                || unit_props2.main_chain_index == None)
+    {
+        let (earlier_unit, later_unit, result_if_found) = if unit_props1.level < unit_props2.level {
+            (unit_props1, unit_props2, -1)
+        } else {
+            (unit_props2, unit_props1, 1)
+        };
+
+        // In JS it is okay if main_chain_index === null, the delta will be negative, need to check
+        // the None value in rust
+        let earlier_unit_delta =
+            earlier_unit.main_chain_index.unwrap() - earlier_unit.latest_included_mc_index.unwrap();
+        let later_unit_delta =
+            later_unit.main_chain_index.unwrap() - later_unit.latest_included_mc_index.unwrap();
```

Comment:
if main_chain_index is None, the unwrap would always panic, which is not the correct behavior, we should avoid panic here, return an Err if possible.